### ISSUE TITLE
[Help wanted] Not satisfaying image_stylization result

### ISSUE THREAD

**Issue by ema987** [tensorflow/magenta#1385]
Hello,

based on the info at https://github.com/tensorflow/magenta/tree/master/magenta/models/image_stylization I tried to train new models from scratch but when transferring the style I get bad results, probably due to the same cause as #935 and #1285.

I report here the more information I can to help solve the issue:

1. I ran
```
image_stylization_create_dataset \
      --vgg_checkpoint=vgg/vgg_16.ckpt \
      --style_files=STYLE_IMAGE.jpg \
      --output_file=TRAIN_RECORD.tfrecord
```

then

```
image_stylization_train \
      --train_dir=TRAIN_DIR \
      --style_dataset_file=TRAIN_RECORD.tfrecord \
      --num_styles=1 \
      --vgg_checkpoint=vgg/vgg_16.ckpt \
      --imagenet_data_dir=imagenet_tf_output
```

After the model was generated I used the following code to transfer the style.

```
image_stylization_transform \
      --num_styles=1 \
      --checkpoint=MODEL.CKPT-XXXX \
      --input_image=INPUT_IMAGE.jpg \
      --which_styles="[0]" \
      --output_dir=OUTPUT_DIR \
      --output_basename="stylized"
```

2. Attached you can find the style image I used for training, starry_night_1280.jpg, resolution is 1280 × 1014 (**I get OOM issues during image_stylization_create_dataset if I use bigger resolution images**)

![starry_night_1280](https://user-images.githubusercontent.com/9298564/50098118-98a78a00-021b-11e9-8ed2-0b3f083a3864.jpg)

3. I used the **default content/style loss hyperparameters** at the one in the scripts, I didn't change them.

4. Attached you can find the content image I used, home_1600.jpg, resolution is 1600 × 992.

![home_1600](https://user-images.githubusercontent.com/9298564/50098151-a3621f00-021b-11e9-93f3-212885269560.jpg)

5. Attached you can find first the style transfer result obtained using the Varied model provided, which has the Starry night style, and then the result I got using the model I trained. If you look at the one obtained with the Varied model (home_varied_model_result.jpg) you can clearly see the starry night style has been applied; if you look at the one obtained using the model I trained (home_custom_model_result.jpg), you can see there is "something" missing.

![home_varied_model_result](https://user-images.githubusercontent.com/9298564/50098218-c7bdfb80-021b-11e9-8032-40512f03a65c.png)

![home_custom_model_result](https://user-images.githubusercontent.com/9298564/50098229-cc82af80-021b-11e9-8fe4-29138c2a4f67.png)

6. Environment info:
  TensorFlow: v1.12.0
  Nvidia K80
  Nvidia driver: v390.77
  Cuda: v9.0
  Installed using conda package https://anaconda.org/anaconda/tensorflow-gpu

Could you please point me in the right direction to discover what's wrong in the process I follow?
Thank you very much.

**Comment by ema987** [comment]
I made another test and I want to report here the results: same flow as above, I just changed the resolution of the style image used to create the model.

I have used the following one, which has a resolution of 3000 × 2375px. I've tried with a 4000px one but I get OOM errors when I run the model training.

![starry_night_3000](https://user-images.githubusercontent.com/9298564/50278837-72b2fd00-0448-11e9-8c70-83e11984a3b1.jpg)

This is the image with the style applied which I got as a result.

![stylized_starry3000_0](https://user-images.githubusercontent.com/9298564/50278879-86f6fa00-0448-11e9-9dc4-69e7c2d56f43.png)

Unfortunately, also this image doesn't show any sign of the starry night style as the one created with the varied model.

**Comment by vdumoulin** [comment]
Hi Emanuele!

The resolution of the content and style images you are using is probably too large for the model. We usually train on square content images of size 256px (see [here](https://github.com/tensorflow/magenta/blob/master/magenta/models/image_stylization/image_stylization_train.py#L40)). I don't recall what exact style image sizes we used, but I remember them being smaller than the resolutions you trained at (this is also the case for other commonly-used fast stylization models; see [here](https://github.com/tensorflow/magenta/blob/master/magenta/models/arbitrary_image_stylization/arbitrary_image_stylization_train.py#L90) and [here](https://github.com/jcjohnson/fast-neural-style/blob/master/scripts/make_style_dataset.py#L58)).

This has several consequences which could explain your observations:
* The VGG network which is used to compute the style transfer loss was trained on 224x224 inputs, and feeding in much larger inputs runs the risk of getting OOM errors. Likewise, the style transfer network architecture proposed by Johnson et al. and adopted for this model was built with 256x256 inputs in mind.
* The Gram matrices computed from the VGG feature maps are not input-scale invariant, meaning that feeding in the same style image at different resolutions will produce different Gram matrices, and consequently will lead to the style loss emphasizing different kinds of visual textures.
* Once the style transfer network is trained, there's an inherent spatial scale to the visual textures that get embedded onto the stylized image. Given that the model is trained with 256x256 content images, feeding a much larger content image at evaluation time will result in the brush strokes and "swirls" appearing comparatively smaller (this is visible in the image you included in point 5). Also, note that stylizing large content images is non-trivial, and this is true even for the optimization-based procedure (see [this paper](https://arxiv.org/abs/1611.07865) for more details).

With all of that in mind, I would suggest that you lower the resolution of the style image you use to train as well as the resolution of the content image you use for evaluation.

Please don't hesitate to reach out again if you have further questions!

**Comment by ema987** [comment]
Hi vdumoulin,

I really appreciate all the information you gave me and the time you spent giving all these insights!
I'll study them and make other attempts, then I'll let you know!

Thank you very much indeed!

**Comment by adarob** [comment]
Please reopen if you have more questions!

**Comment by ema987** [comment]
Hi @adarob , could you please reopen the ticket? I have got time to work on it only in the last few days, that's why I didn't reply earlier.

@vdumoulin I followed your suggestions and I trained a new model from a 256x256px starry night image, the following one.

![starry_night_256_256](https://user-images.githubusercontent.com/9298564/50893886-3e58ae00-1402-11e9-8117-84a341307b88.jpg)

After that, I tried to stylize the same image I used in the previous attempts, but actually not much seems to be changed. Here the result.

![starry256_1600](https://user-images.githubusercontent.com/9298564/50893964-647e4e00-1402-11e9-81ff-02f7491954f9.png)

Even using smaller content images the transfer result looks like this one. A smaller one attached.

![starry256_480](https://user-images.githubusercontent.com/9298564/50894057-94c5ec80-1402-11e9-84b2-a5e6f84ec410.png)

--------

I tried using another content image, the following:

![japan-panorama-12-min](https://user-images.githubusercontent.com/9298564/50894148-cccd2f80-1402-11e9-8644-b87304b529d1.jpg)

And I transferred the style of my model:

![starry256_japan_0](https://user-images.githubusercontent.com/9298564/50894229-fb4b0a80-1402-11e9-8908-b51565abb6a3.png)

and of yours:

![varied_japan_22](https://user-images.githubusercontent.com/9298564/50894237-01d98200-1403-11e9-9e6f-128dbfb97a6b.png)

Here the result looks much better than the one transferred to the previous content image (the house) but still I see imperfections and I see your model does a better job, so actually I wasn't able to recreate one similar to yours.

Any further suggestion you have will be very appreciated. Thank you for your time.
